Require email confirmation before "borrowing" items

crismali opened the pull request:
# What it does

This adds a check to prevent users from placing holds on items before confirming their email address (via devise). It also adds some warning toast messages to encourage users to confirm.

# Why it is important

Handles #1370 

# UI Change Screenshot

Verify email message on the search page:
![Screenshot 2024-08-07 at 3 21 06 PM](https://github.com/user-attachments/assets/c678d9d1-b93f-47e2-901d-7e68831753bb)

Verify email message on the item page:
![Screenshot 2024-08-07 at 3 20 58 PM](https://github.com/user-attachments/assets/d57a5354-d413-464e-a950-bc969b47ea14)

Message where a user would normally see the "Place Hold" button:
![Screenshot 2024-08-07 at 3 21 51 PM](https://github.com/user-attachments/assets/caab7e82-cb99-4d28-bb9d-cb8abe61395c)

Member login page:
![Screenshot_2024-08-07_at_3_22_07 PM](https://github.com/user-attachments/assets/455ae56c-9b07-4e0f-9524-934014c70feb)

Resend confirmation instructions page:
![Screenshot 2024-08-07 at 3 22 18 PM](https://github.com/user-attachments/assets/468ccde4-9599-4930-9f41-5073e30b2a1b)

# Implementation notes

It doesn't look like members can create loans directly, so I prevented members with unconfirmed email addresses from being able to make holds. Admins could still create loans for these members, so maybe we want to add a message on the admin side that warns them about creating loans for these members?

Right now the migration will mark all existing users as verified so these changes should only impact new members. I'm not sure if there's a heuristic we could use to make this smarter.

Take a look and let me know what you think!

crismali commented on app/views/devise/confirmations/new.html.erb:
```diff
@@ -1,16 +1,16 @@
+      <%= f.submit "Resend Confirmation Instructions", class: "btn btn-primary" %>
+    </div>
+  <% end %>
+</div>
```
Changed this template to look more like the other devise pages.

jim replied:
Good catch!

crismali commented on config/initializers/devise.rb:
```diff
@@ -134,7 +134,7 @@
   # Default is 0.days, meaning the user cannot access the website without
   # confirming their account.
-  # config.allow_unconfirmed_access_for = 2.days
+  config.allow_unconfirmed_access_for = 2.days
```
2 days is the default here, but maybe it should be more like a week?

phinze replied:
Since we're explicitly disallowing the thing we don't want people to do without confirming, I wonder if we should set it to [`nil` so it never kicks them out?](https://github.com/heartcombo/devise/pull/2278)

jim replied:
I agree that if we're going to go with the model of preventing actions within the account based on if an email is confirmed, it's OK to always let folks log in.

phinze commented:
This is looking really good!

> Admins could still create loans for these members, so maybe we want to add a message on the admin side that warns them about creating loans for these members?

I think something on the admin side member page to indicate that the user has not confirmed their email address is a good idea.

phinze commented:
I'm flushing out a few Circulate tasks now, so I can pick up the two outstanding tweaks and bring this home. 🤝

phinze commented:
After discussing this with staff, we have two changes to make here:

 * Prevent users from logging in when verification is incomplete
 * Change message copy to use active voice e.g. "You have not verified your email address yet! Please click the confirmation link in the email you received from us. Click here to request a new email."

And one change to investigate:

 * Can we consolidate the confirmation link in with the new member email?

crismali commented:
I've made some of the changes we talked about the other night. A new question is how much do we want to surface folks who have changed their email addresses but haven't confirmed the new ones yet. I've changed the members so that when they change the member email they're also updated the user unconfirmed email so there shouldn't be in any place where a user's email being changed directly.  I've also made it so the message on the admin member page appears if the user has never been confirmed or if they just haven't confirmed their new email. 

What do folks think? Should we call this out more for members or admins?

phinze commented on app/views/devise/sessions/new.html.erb:
```diff
@@ -37,6 +37,7 @@
         <li>If you don't have a membership with us yet, <%= link_to "sign up online", signup_path %>.</li>
       <% end %>
         <li>If you need to choose or don't remember your password, <%= link_to "reset your password", new_user_password_path %>.</li>
+        <li>If you need haven't confirmed your email address, <%= link_to "request a confirmation email", new_user_confirmation_path %>.</li>
```
```suggestion
        <li>If you haven't confirmed your email address, <%= link_to "request a confirmation email", new_user_confirmation_path %>.</li>
```

Found a typo, though I think it might be more confusing to have this showing to every logging in user vs only showing it in flash messages when the email is not verified.

phinze replied:
It looks like we can modify the devise settings so the flash message includes a link instead. https://stackoverflow.com/questions/4008059/rails-devise-customize-flash-message-devise-en-yml-with-a-link-to

jim reviewed (approved):
I think this is looking good. I just want to think through if we need to do any communication before merging and deploying, since every existing user will suddenly not be able to log in.

crismali commented:
@jim They should all be able to log in with this. The migration makes it so that if you're already an existing user you're automatically confirmed even though that hasn't actually happened.
